### PR TITLE
Permutation handedness

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: geometry/permutation_body.hpp [hunk 1]
```diff
@@ -144,74 +154,42 @@ template<typename FromFrame, typename ThroughFrame, typename ToFrame>
...
+      /*ZXY*/ {E::ZXY, E::XYZ, E::YZX, O::YXZ, O::XZY, O::ZYX},
+      /*XZY*/ {O::XZY, O::YXZ, O::ZYX, E::XYZ, E::ZXY, E::YZX},
+      /*ZYX*/ {O::ZYX, O::XZY, O::YXZ, E::YZX, E::XYZ, E::ZXY},
+      /*YXZ*/ {O::YXZ, O::ZYX, O::XZY, E::ZXY, E::YZX, E::XYZ},
+  }};
+
+  return Result(static_cast<Result::CoordinatePermutation>(
+      multiplication
+          [0b111 & (static_cast<int>(left.coordinate_permutation_) >> INDEX)]
+          [0b111 & (static_cast<int>(right.coordinate_permutation_) >> INDEX)]
```
Comment (pleroy) [inline geometry/permutation_body.hpp]: Name the masks.

File: geometry/permutation_body.hpp [hunk 2]
```diff
@@ -144,74 +154,42 @@ template<typename FromFrame, typename ThroughFrame, typename ToFrame>
...
-                      {PFT::YZX, "YZX"},
-                      {PFT::ZXY, "ZXY"},
-                      {PFT::XZY, "XZY"},
-                      {PFT::ZYX, "ZYX"},
-                      {PFT::YXZ, "YXZ"}};
-  return out << debug_string.at(permutation.coordinate_permutation_);
+  static constexpr std::array<std::string_view, 6> debug_string{
+      {"XYZ", "YZX", "ZXY", "XZY", "ZYX", "YXZ"}};
+  return out
+         << debug_string[0b111 & (static_cast<int>(
```
Comment (pleroy) [inline geometry/permutation_body.hpp]: Name the mask.

File: geometry/rotation_test.cpp [hunk 3]
```diff
@@ -321,29 +321,37 @@ TEST_F(RotationTest, Enums) {
   // given |permutation|, and with appropriate sign changes, is equivalent to
   // the ZXZ convention.
   auto const check_euler_angles = [&α, &β, &γ, &zxz_euler, this](
```
Comment (pleroy) [inline geometry/rotation_test.cpp]: Capture `this` first.

File: geometry/rotation_test.cpp [hunk 4]
```diff
@@ -321,29 +321,37 @@ TEST_F(RotationTest, Enums) {
...
-      EulerAngles axes,
-      Permutation<World, World>::CoordinatePermutation permutation) {
-    Permutation<World, World> const σ(permutation);
-    Permutation<World1, World1> const τ =
-        (Permutation<World, World1>::Identity() * σ *
+      EulerAngles axes, auto permutation) {
+    auto const permuted_handedness =
```
Comment (pleroy) [inline geometry/rotation_test.cpp]: `constexpr`?

File: geometry/rotation_test.cpp [hunk 5]
```diff
@@ -353,29 +361,38 @@ TEST_F(RotationTest, Enums) {
...
-    Permutation<World1, World1> const τ =
-        (Permutation<World, World1>::Identity() * σ *
-         Permutation<World1, World>::Identity()).Inverse();
-    Rotation<World, World1> const cardano(σ.Determinant() * α,
-                                          σ.Determinant() * β,
-                                          σ.Determinant() * γ,
-                                          axes,
-                                          DefinesFrame<World1>{});
+                                        CardanoAngles axes, auto permutation) {
+    auto const permuted_handedness =
```
Comment (pleroy) [inline geometry/rotation_test.cpp]: Same.

File: geometry/rotation_test.cpp [hunk 6]
```diff
@@ -353,29 +361,38 @@ TEST_F(RotationTest, Enums) {
...
+            : Handedness::Left;
+    using Permuted =
+        Frame<enum class PermutedTag, Inertial, permuted_handedness>;
+    using PermutedRotated =
+        Frame<enum class PermutedRotatedTag, Inertial, permuted_handedness>;
+    Permutation<World, Permuted> const σ(permutation);
+    Permutation<PermutedRotated, World1> const τ =
+        (Permutation<Permuted, PermutedRotated>::Identity() * σ *
+         Permutation<World1, World>::Identity())
+            .Inverse();
```
Comment (pleroy) [inline geometry/rotation_test.cpp]: Move to previous line.

File: physics/rigid_motion_test.cpp [hunk 7]
```diff
@@ -120,8 +121,8 @@ class RigidMotionTest : public testing::Test {
           RigidTransformation<Geocentric, Selenocentric>(
               Geocentric::origin + earth_to_moon_,
               Selenocentric::origin,
-              Permutation<Geocentric, Selenocentric>(
-                  Permutation<Geocentric, Selenocentric>::YXZ).Forget()),
+              Permutation<Geocentric, Selenocentric>(EvenPermutation::YXZ)
+                  .Forget()),
```
Comment (pleroy) [inline physics/rigid_motion_test.cpp]: Might fit on the previous line.

File: ksp_plugin/plugin.cpp [hunk 8]
```diff
@@ -313,8 +314,7 @@ Rotation<BodyWorld, World> Plugin::CelestialRotation(
   // |BodyWorld| with its y and z axes swapped (so that z is the polar axis).
   // The basis is right-handed.
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Remove this line.

File: geometry/permutation_test.cpp [hunk 9]
```diff
@@ -103,135 +105,185 @@ TEST_F(PermutationTest, Determinant) {
...
+  auto const test_composition_for = [&](auto const& lhs,
+                                        auto const& rhs,
+                                        auto const from_vector) {
+    for (auto const& left : lhs) {
+      for (auto const& right : rhs) {
+        auto const composition = left * right;
+        auto const composition_as_orthogonal_maps =
+            left.Forget() * right.Forget();
+        for (Length l = 1 * Metre; l < 4 * Metre; l += 1 * Metre) {
+          // In C++20 we could have template parameters on the lambda which
```
Comment (pleroy) [inline geometry/permutation_test.cpp]: Make this comment a TODO.

File: geometry/permutation_test.cpp [hunk 10]
```diff
@@ -103,135 +105,185 @@ TEST_F(PermutationTest, Determinant) {
...
 TEST_F(PermutationDeathTest, SerializationError) {
-  Identity<World1, World2> id;
-  EXPECT_DEATH({
-    serialization::LinearMap message;
-    id.WriteToMessage(&message);
-    Perm const p = Perm::ReadFromMessage(message);
-  }, "HasExtension.*Permutation");
+  Identity<R1, R2> id;
+  EXPECT_DEATH(
+      {
```
Comment (pleroy) [inline geometry/permutation_test.cpp]: Keep the old formatting, it was more compact.

File: physics/rigid_motion_test.cpp (hunk 7 above)
Comment (eggrobin) [reply physics/rigid_motion_test.cpp]: No.

**Pull request merged**